File: ruby/lib/testivai/capture.rb
# frozen_string_literal: true

require "base64"
require "json"
require "fileutils"
require "pathname"

require_relative "config"
require_relative "element_map"

module Testivai
  # Capture half of the pipeline: writes `.testivai/temp/<name>/` with the
  # screenshot, the DOM snapshot, and the element map. The compare half is
  # `npx testivai report`, shared with every other adapter.
  #
  # Works against any Capybara session or a bare Selenium driver — all this
  # needs is `execute_script` plus some way to take a screenshot, which is
  # exactly what every Capybara driver already exposes.
  module Capture
    STYLE_ID = "__testivai_capture_style__"

    # Collapses animations and transitions to ~0s, hides the caret, and
    # disables smooth scrolling — the top causes of flaky visual diffs.
    # Near-zero rather than `none` so animations land on their final frame
    # instead of never rendering.
    STABILIZE_CSS = <<~CSS
      *, *::before, *::after {
        animation-duration: 0.001s !important;
        animation-delay: 0s !important;
        animation-iteration-count: 1 !important;
        transition-duration: 0.001s !important;
        transition-delay: 0s !important;
        caret-color: transparent !important;
        scroll-behavior: auto !important;
      }
    CSS

    INJECT_STYLE_JS = <<~JS
      var s = document.createElement('style');
      s.id = arguments[0];
      s.textContent = arguments[1];
      document.head.appendChild(s);
    JS

    REMOVE_STYLE_JS = <<~JS
      var s = document.getElementById(arguments[0]);
      if (s && s.parentNode) { s.parentNode.removeChild(s); }
    JS

    FONTS_READY_JS = <<~JS
      return !document.fonts || document.fonts.status === 'loaded';
    JS

    DOM_SNAPSHOT_JS = <<~JS
      var clone = document.documentElement.cloneNode(true);
      var sels = arguments[0] || [];
      for (var i = 0; i < sels.length; i++) {
        try {
          var found = clone.querySelectorAll(sels[i]);
          for (var j = 0; j < found.length; j++) { found[j].remove(); }
        } catch (e) {}
      }
      return clone.outerHTML;
    JS

    module_function

    # Capture a snapshot.
    #
    # @param session  a Capybara session (`page`) or a Selenium driver
    # @param name     snapshot name; becomes the directory under temp/
    # @return [Pathname] the temp directory written
    def witness(session, name,
                ignore_selectors: nil,
                stabilize: nil,
                skip_dom: false,
                skip_element_map: false,
                max_elements: nil,
                project_root: nil)
      browser = resolve_browser(session)
      root = Pathname.new(project_root || Config.project_root)
      config = Config.load(root)

      selectors = ((config["ignoreSelectors"] || []) + (ignore_selectors || [])).uniq
      stabilize = config["stabilize"] if stabilize.nil?
      stabilize = true if stabilize.nil?

      temp_dir = root.join(".testivai", "temp", name.to_s)
      FileUtils.mkdir_p(temp_dir)

      css = +""
      css << STABILIZE_CSS if stabilize
      css << hide_css(selectors) unless selectors.empty?

      injected = false
      unless css.empty?
        injected = inject_css(browser, css)
        wait_for_fonts(browser) if stabilize
      end

      begin
        screenshot = capture_screenshot(browser)
      ensure
        remove_css(browser) if injected
      end

      File.binwrite(temp_dir.join("screenshot.png"), screenshot)

      # DOM snapshot — best-effort. Losing it only costs the noise hint;
      # it must never break the screenshot path.
      unless skip_dom
        begin
          dom = browser.execute_script(DOM_SNAPSHOT_JS, selectors)
          temp_dir.join("dom.html").write(dom) if dom.is_a?(String) && !dom.empty?
        rescue StandardError # rubocop:disable Lint/SuppressedException
        end
      end

      # Element map — same best-effort contract. Powers region→selector
      # attribution, shift classification, and the style fingerprint.
      unless skip_element_map
        begin
          expr = ElementMap.expression(max_elements || ElementMap::DEFAULT_MAX_ELEMENTS, selectors)
          map = browser.execute_script(expr)
          if map.is_a?(Array) && !map.empty?
            temp_dir.join("elements.json").write(JSON.generate(map))
          end
        rescue StandardError # rubocop:disable Lint/SuppressedException
        end
      end

      temp_dir
    end

    # Capybara sessions wrap the real driver; a bare Selenium driver is
    # already what we want. Accept both so this works in RSpec/Capybara,
    # Cucumber, or plain selenium-webdriver.
    def resolve_browser(session)
      return session.driver.browser if session.respond_to?(:driver) && session.driver.respond_to?(:browser)
      return session.browser if session.respond_to?(:browser)

      session
    end

    def hide_css(selectors)
      selectors.map { |s| "#{s} { visibility: hidden !important; }" }.join("\n")
    end

    def inject_css(browser, css)
      browser.execute_script(INJECT_STYLE_JS, STYLE_ID, css)
      true
    rescue StandardError
      false
    end

    def remove_css(browser)
      browser.execute_script(REMOVE_STYLE_JS, STYLE_ID)
    rescue StandardError # rubocop:disable Lint/SuppressedException
    end

    def wait_for_fonts(browser, timeout: 10.0)
      deadline = Time.now + timeout
      loop do
        return if browser.execute_script(FONTS_READY_JS)
        return if Time.now > deadline

        sleep 0.05
      end
    rescue StandardError # rubocop:disable Lint/SuppressedException
    end

    # Full page where the browser offers it, viewport otherwise.
    #
    #   - Chromium: CDP Page.captureScreenshot with captureBeyondViewport,
    #     the same mechanism Playwright uses — a true full-page capture with
    #     no window resizing (resizing breaks 100vh layouts).
    #   - Firefox: selenium-webdriver exposes save_full_page_screenshot.
    #   - anything else: the plain viewport screenshot.
    def capture_screenshot(browser)
      if browser.respond_to?(:execute_cdp)
        begin
          result = browser.execute_cdp("Page.captureScreenshot",
                                       "captureBeyondViewport" => true, "format" => "png")
          data = result && (result["data"] || result[:data])
          return Base64.decode64(data) if data
        rescue StandardError # rubocop:disable Lint/SuppressedException
        end
      end

      if browser.respond_to?(:full_screenshot_as)
        begin
          return browser.full_screenshot_as(:png)
        rescue StandardError # rubocop:disable Lint/SuppressedException
        end
      end

      browser.screenshot_as(:png)
    end
  end
end

File: python/src/testivai/_capture.py
"""
Capture implementation for playwright-python (sync API).

Semantics mirror the JS adapters exactly (@testivai/witness-playwright):
  1. stabilization CSS: animations/transitions complete instantly at their
     FINAL state, caret hidden, smooth scroll off; wait for web fonts
  2. ignoreSelectors: hidden from pixels (visibility:hidden) AND excluded
     from the DOM snapshot (one consistent semantic)
  3. variant keying: `<name>__<variant>` so parallel browsers/viewports
     never overwrite each other's baselines
  4. on-disk contract: .testivai/temp/<name>/{screenshot.png, dom.html}

The compare/report half stays in @testivai/witness — run `testivai report`
(see runner.run_report), which playwright-python users can always do because
Playwright for Python ships with a Node.js driver.
"""

from __future__ import annotations

import json
import re
import time
from importlib import resources
from pathlib import Path
from typing import Any, Optional, Sequence

#: Mirrors STABILIZE_CSS in @testivai/witness-playwright (config/stabilize.ts).
#: Near-zero durations (not `none`) let entry animations COMPLETE at their
#: final state — pages whose content starts hidden render fully.
STABILIZE_CSS = (
    "*, *::before, *::after {"
    " animation-duration: 0.001s !important;"
    " animation-delay: 0s !important;"
    " animation-iteration-count: 1 !important;"
    " transition-duration: 0.001s !important;"
    " transition-delay: 0s !important;"
    " caret-color: transparent !important;"
    " scroll-behavior: auto !important;"
    " }"
)

_FONTS_READY_JS = "document.fonts ? document.fonts.status !== 'loading' : true"

_DOM_SNAPSHOT_JS = """
(selectors) => {
  const clone = document.documentElement.cloneNode(true);
  for (const sel of selectors) {
    try { clone.querySelectorAll(sel).forEach(el => el.remove()); } catch (e) {}
  }
  return clone.outerHTML;
}
"""


def sanitize_variant(variant: str) -> str:
    """Match the JS adapters: non [a-z0-9_-] runs become '_', lowercased."""
    return re.sub(r"[^a-z0-9_-]+", "_", variant, flags=re.IGNORECASE).lower()


def load_local_config(project_root: Path) -> dict:
    """Read .testivai/config.json; defaults mirror @testivai/witness."""
    defaults: dict[str, Any] = {
        "mode": "local",
        "threshold": 0.1,
        "autoOpen": True,
        "failOnDiff": False,
        "stabilize": True,
        "ignoreSelectors": [],
    }
    config_path = project_root / ".testivai" / "config.json"
    if not config_path.exists():
        return defaults
    try:
        loaded = json.loads(config_path.read_text())
        return {**defaults, **loaded}
    except (OSError, json.JSONDecodeError):
        return defaults


def build_ignore_css(selectors: Sequence[str]) -> str:
    """visibility:hidden preserves layout — same rule as the JS adapters."""
    return "\n".join(f"{sel} {{ visibility: hidden !important; }}" for sel in selectors)


def witness(
    page: Any,
    name: str,
    *,
    ignore_selectors: Optional[Sequence[str]] = None,
    stabilize: Optional[bool] = None,
    variant: Optional[str] = None,
    skip_dom: bool = False,
    project_root: Optional[Path] = None,
) -> Path:
    """
    Capture a visual snapshot from a playwright-python (sync API) Page.

    Writes `.testivai/temp/<name>/screenshot.png` (+ `dom.html`), then
    `testivai report` (or the pytest plugin's session hook) compares against
    committed baselines.

    Args:
        page: playwright.sync_api.Page (duck-typed: screenshot / evaluate /
            add_style_tag are all that's required).
        name: Snapshot name; becomes the directory under .testivai/temp/.
        ignore_selectors: Per-call CSS selectors hidden from pixels and
            excluded from the DOM snapshot. Merged with the global
            `ignoreSelectors` from .testivai/config.json.
        stabilize: Override the stabilization default (config `stabilize`,
            default True).
        variant: Multi-browser/viewport key, folded into the name as
            `<name>__<variant>` so parallel runs never collide.
        skip_dom: Skip the DOM snapshot (disables the noise hint for this
            snapshot; pixel diff still works).
        project_root: Defaults to the current working directory.

    Returns:
        The temp directory the capture was written to.
    """
    if not name or not isinstance(name, str):
        raise ValueError("testivai.witness: snapshot name is required and must be a non-empty string")

    root = Path(project_root) if project_root else Path.cwd()
    config = load_local_config(root)

    effective_stabilize = config["stabilize"] if stabilize is None else stabilize
    merged_selectors = list(dict.fromkeys([*config.get("ignoreSelectors", []), *(ignore_selectors or [])]))

    if variant:
        name = f"{name}__{sanitize_variant(variant)}"

    # 1. Prepare the page: one style tag covering stabilization + ignores
    css_parts = []
    if effective_stabilize:
        css_parts.append(STABILIZE_CSS)
    if merged_selectors:
        css_parts.append(build_ignore_css(merged_selectors))

    style_handle = None
    if css_parts:
        try:
            style_handle = page.add_style_tag(content="\n".join(css_parts))
        except Exception:
            style_handle = None  # locked-down page; capture proceeds
        if effective_stabilize:
            _wait_for_fonts(page)

    temp_dir = root / ".testivai" / "temp" / name
    temp_dir.mkdir(parents=True, exist_ok=True)

    try:
        # 2. Full-page screenshot via Playwright's native API
        page.screenshot(path=str(temp_dir / "screenshot.png"), full_page=True)
    finally:
        if style_handle is not None:
            try:
                style_handle.evaluate("el => el.remove()")
            except Exception:
                pass  # best-effort cleanup

    # 3. DOM snapshot with ignored elements removed (best-effort — a flaky
    #    page never breaks the screenshot path; missing dom.html just
    #    suppresses the noise hint)
    if not skip_dom:
        try:
            dom = page.evaluate(_DOM_SNAPSHOT_JS, list(merged_selectors))
            if isinstance(dom, str) and dom:
                (temp_dir / "dom.html").write_text(dom, encoding="utf-8")
        except Exception:
            pass

    return temp_dir


def _wait_for_fonts(page: Any, timeout_seconds: float = 10.0) -> None:
    """Bounded wait for web fonts — a fallback-font capture diffs 30%+."""
    deadline = time.monotonic() + timeout_seconds
    while time.monotonic() < deadline:
        try:
            if page.evaluate(_FONTS_READY_JS):
                return
        except Exception:
            return
        time.sleep(0.1)


# ── Element-map collector ──────────────────────────────────────────────────
#
# `element_map.js` is GENERATED from packages/witness/src/capture/element-map.ts
# by scripts/generate-element-map-asset.js, and CI fails if it is stale. Every
# adapter -- TypeScript, Python, Java -- injects that identical function, which
# matters because all of them write into one shared `.testivai/baselines/`
# directory: two languages producing subtly different maps for the same page
# would surface as phantom regressions.

_DEFAULT_MAX_ELEMENTS = 3000

_ELEMENT_MAP_SRC: Optional[str] = None


def _load_element_map_source() -> str:
    """
    Read the generated collector, cached after the first call.

    The file starts with a "do not edit" banner comment that is useful to a
    reader but pure weight on the wire, so it is sliced off: every capture
    ships this string to the browser.
    """
    global _ELEMENT_MAP_SRC
    if _ELEMENT_MAP_SRC is None:
        raw = resources.files("testivai").joinpath("element_map.js").read_text(
            encoding="utf-8"
        )
        start = raw.find("function collectElementMap")
        if start == -1:
            raise RuntimeError(
                "element_map.js is malformed: collectElementMap not found. "
                "Regenerate with scripts/generate-element-map-asset.js"
            )
        _ELEMENT_MAP_SRC = raw[start:].strip()
    return _ELEMENT_MAP_SRC


def _element_map_expression(max_elements: int, ignore_selectors: Sequence[str]) -> str:
    """
    Wrap the collector exactly as `buildElementMapExpression` does on the
    TypeScript side. Selenium/WebDriver needs the explicit `return`.
    """
    return "return ({0})(document, window, {1}, {2});".format(
        _load_element_map_source(),
        int(max_elements),
        json.dumps(list(ignore_selectors)),
    )


# ── Page-settled probe ──────────────────────────────────────────────────────
#
# `settle.js` is GENERATED from packages/witness/src/capture/settle.ts, same as
# element_map.js, so every language polls the identical predicate.
#
# Deliberately NOT network idle: Playwright's own docs mark that DISCOURAGED for
# testing, and it is the wrong signal for a visual snapshot — a page with
# analytics beacons never goes quiet, while a network-idle page can still be
# animating. What matters is whether the rendered page stopped changing.

DEFAULT_QUIET_MS = 150
DEFAULT_SETTLE_TIMEOUT = 5.0

_SETTLE_SRC: Optional[str] = None


def _load_settle_source() -> str:
    global _SETTLE_SRC
    if _SETTLE_SRC is None:
        raw = resources.files("testivai").joinpath("settle.js").read_text(encoding="utf-8")
        start = raw.find("function settleProbe")
        if start == -1:
            raise RuntimeError(
                "settle.js is malformed: settleProbe not found. "
                "Regenerate with scripts/generate-element-map-asset.js"
            )
        _SETTLE_SRC = raw[start:].strip()
    return _SETTLE_SRC


def _settle_expression(quiet_ms: int) -> str:
    return "return ({0})(document, window, {1});".format(_load_settle_source(), int(quiet_ms))


def wait_for_settled(driver: Any, quiet_ms: int = DEFAULT_QUIET_MS,
                     timeout: float = DEFAULT_SETTLE_TIMEOUT) -> None:
    """
    Poll until the page stops changing, or give up.

    Always bounded: a page that never settles yields a capture, not a hang. A
    driver that cannot evaluate the probe returns something unusable, in which
    case we proceed at once rather than paying the whole timeout every capture.
    """
    expr = _settle_expression(quiet_ms)
    deadline = time.time() + timeout
    while True:
        try:
            state = driver.execute_script(expr)
        except Exception:
            return
        if not isinstance(state, dict) or not isinstance(state.get("settled"), bool):
            return
        if state["settled"] or time.time() >= deadline:
            return
        time.sleep(0.05)


def stop_settle_observer(driver: Any) -> None:
    """Detach the observer so it does not linger on the page under test."""
    try:
        driver.execute_script(
            "try { var s = window.__testivaiSettleState;"
            " if (s && s.observer && s.observer.disconnect) s.observer.disconnect(); } catch (e) {}"
            " try { delete window.__testivaiSettleState; } catch (e) {}"
        )
    except Exception:
        pass
